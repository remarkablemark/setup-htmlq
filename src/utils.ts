import os from 'os';
import path from 'path';

const platform = {
  darwin: 'darwin',
  linux: 'linux',
  win32: 'windows',
} as const;

/**
 * Gets a string identifying the operating system platform.
 *
 * @see {@link https://nodejs.org/api/os.html#os_os_platform}
 *
 * @param os - OS in [darwin, linux, win32...]
 * @returns - Return value in [darwin, linux, windows]
 */
function getOS(os: NodeJS.Platform) {
  return platform[os as keyof typeof platform];
}

/**
 * Gets download URL.
 *
 * @see {@link https://github.com/mgdm/htmlq/releases}
 *
 * @param version - CLI version
 * @param name - CLI name
 * @returns - Download URL
 */
export function getDownloadUrl(version: string) {
  const platform = os.platform();

  const filename = `htmlq-x86_64-${getOS(platform)}`;
  const extension = platform === 'win32' ? 'zip' : 'tar.gz';

  return `https://github.com/mgdm/htmlq/releases/download/v${version}/${filename}.${extension}`;
}

/**
 * Gets CLI path.
 *
 * @param directory - Directory
 * @param name - CLI name
 * @returns - Binary path
 */
export function getFilepath(directory: string, name: string) {
  return path.join(directory, name + (os.platform() === 'win32' ? '.exe' : ''));
}
