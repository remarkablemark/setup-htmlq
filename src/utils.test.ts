import os from 'os';

import { getDownloadUrl, getFilepath } from './utils';

jest.mock('os');

const mockedOs = jest.mocked(os);

const platforms = ['darwin', 'linux', 'win32'] as const;

describe('getDownloadUrl', () => {
  describe.each(platforms)('when OS is %p', (os) => {
    const { RUNNER_TEMP } = process.env;
    const version = '0.4.0';

    beforeAll(() => {
      process.env.RUNNER_TEMP = '';
    });

    afterAll(() => {
      process.env.RUNNER_TEMP = RUNNER_TEMP;
    });

    beforeEach(() => {
      jest.resetAllMocks();
      mockedOs.platform.mockReturnValueOnce(os);
      mockedOs.tmpdir.mockReturnValueOnce('/var/folders/1/tmp/T');
    });

    it('gets download object', () => {
      expect(getDownloadUrl(version)).toMatchSnapshot();
    });
  });
});

describe('getFilepath', () => {
  describe.each(platforms)('when OS is %p', (os) => {
    beforeEach(() => {
      jest.resetAllMocks();
      mockedOs.platform.mockReturnValueOnce(os);
    });

    it('returns CLI filepath', () => {
      const directory = 'directory';
      const name = 'name';
      expect(getFilepath(directory, name)).toMatchSnapshot();
    });
  });
});
