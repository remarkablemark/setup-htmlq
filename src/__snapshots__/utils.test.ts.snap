// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getBinaryPath when OS is "darwin" returns CLI filepath 1`] = `"directory/name"`;

exports[`getBinaryPath when OS is "linux" returns CLI filepath 1`] = `"directory/name"`;

exports[`getBinaryPath when OS is "win32" returns CLI filepath 1`] = `"directory/name.exe"`;

exports[`getDownloadUrl when OS is "darwin" gets download object 1`] = `"https://github.com/mgdm/htmlq/releases/download/v0.4.0/htmlq-x86_64-darwin.tar.gz"`;

exports[`getDownloadUrl when OS is "linux" gets download object 1`] = `"https://github.com/mgdm/htmlq/releases/download/v0.4.0/htmlq-x86_64-linux.tar.gz"`;

exports[`getDownloadUrl when OS is "win32" gets download object 1`] = `"https://github.com/mgdm/htmlq/releases/download/v0.4.0/htmlq-x86_64-windows.zip"`;
